### PR TITLE
Load

### PR DESCRIPTION
**Pull request by yamins81**
Hi @qbilius here is a PR for the code that I wrote last week.  There are a couple of issues I'd like to discuss with you about it.    

Here are some questions -- these are mostly notes for us when we talk:
1) I still need to write a test for the feature extraction API. 
2) do we want to load from __RECENT for validation, or just allow loading from the "real" storage?
3) do we want to force load_params to be used instead of save params? 
4) am I handling scoping correctly?   I have some questions/comments here about exactly what we want to do.

### CODE REVIEW

**Comment by yamins81** [comment]
We agreed to group the saving db params as "save_params" in analogy to load_params

**Comment by yamins81** [comment]
ok @qbilius I've added an extraction test and done the save_params items stuff we agreed to.   Can you have a look at the PR now for review?

**Comment by qbilius** [comment]
Great! I'll get to it on Friday. Tomorrow is nearly the whole day of meetings for me.

File: tfutils/base.py [hunk 1]
```diff
@@ -358,29 +496,31 @@ def run(sess,
...
-             model_params,
-             train_params,
-             loss_params=None,
-             learning_rate_params=None,
-             optimizer_params=None,
-             validation_params=None,
-             thres_loss=100,
-             num_steps=1000000,
-             log_device_placement=False,
+def train_base(save_params,
```
Comment (qbilius) [inline tfutils/base.py]: I don't like the `_base` anymore – too undescriptive. How about `train_from_params`? Or your suggestion.

File: tfutils/base.py [hunk 2]
```diff
@@ -97,83 +98,131 @@ def __init__(self,
...
             os.makedirs(self.cache_dir)
+        
+    def load_rec(self):
+        #first try and see if anything with the save data exists, since obviously
+        #we dont' want to keep loading from the original load location if some work has
+        #already been done
+        load = self.load_from_db({'exp_id': self.exp_id},
+                                 cache_model=True)
+        #if not, try loading from the loading location
+        if not load and not self.sameloc:
```
Comment (yamins81) [inline tfutils/base.py]: hm....  no I think that when no record is found, actually the return is a single None (see line 262

File: tfutils/base.py [hunk 3]
```diff
@@ -97,83 +98,131 @@ def __init__(self,
...
+            load = self.load_from_db(self.load_query,
+                                     cache_model=True,
```
Comment (yamins81) [inline tfutils/base.py]: sure will change

File: tfutils/base.py [hunk 4]
```diff
@@ -97,83 +98,131 @@ def __init__(self,
...
+                                     collfs=self.load_collfs,
+                                     collfs_recent=self.load_collfs_recent)
+            if load is None:
```
Comment (yamins81) [inline tfutils/base.py]: My comment is same as before -- load *will* be None is no record is found.

File: tfutils/base.py [hunk 5]
```diff
@@ -97,83 +98,131 @@ def __init__(self,
...
         """
         # fetch record from database and get the filename info from record
-        if self._restore:
-            load = self.load_from_db({'exp_id': self.exp_id,
-                                      'saved_filters': True},
-                                     cache_model=True,
-                                     force_fetch=force_fetch)
-            if load is not None:
-                rec, cache_filename = load
+        tf_saver = self.tf_saver
```
Comment (yamins81) [inline tfutils/base.py]: Well, I need access to tf_saver below and I didn't want to keep typing "self.tf_saver" ... I guess since it's only used once that's stupid. so I can remove the line.

File: tfutils/base.py [hunk 6]
```diff
@@ -184,26 +233,35 @@ def load_from_db(self, query, cache_model=False, force_fetch=False):
...
             if ckpt_record is None or ckpt_record_recent['uploadDate'] > ckpt_record['uploadDate']:
-                loading_from = self.coll_recent
+                loading_from = coll_recent
                 ckpt_record = ckpt_record_recent
 
         if count + count_recent == 0:  # no matches for query
             log.warning('No matching checkpoint for query "{}"'.format(repr(query)))
             return
 
+        database = loading_from._Collection__database
```
Comment (yamins81) [inline tfutils/base.py]: No because if count + count_recent > 0 then loading_from will not be undefined, since otherwise, the function will return beforehand.

File: tfutils/base.py [hunk 7]
```diff
@@ -232,65 +290,78 @@ def save(self, train_res, valid_res):
...
                'params': self.SONified_params,
                'saved_filters': False,
-               'step': step,
                'duration': duration}
-        if 'optimizer' in train_res:
-            del train_res['optimizer']
-        rec['train_results'] = train_res
+               
+        if train_res:
+            step = self.global_step.eval(session=self.sess)
```
Comment (yamins81) [inline tfutils/base.py]: Yeah, there should be a clearer error message.    I'll do a try/except.

File: tfutils/base.py [hunk 8]
```diff
@@ -515,28 +657,33 @@ def run_base(saver_params,
             model_kwargs['cfg_initial'] = None
         if 'seed' not in model_kwargs:
             model_kwargs['seed'] = 0
-        train_outputs, cfg_final = model_func(inputs=train_inputs,
+            train_outputs, cfg_final = model_func(inputs=train_inputs,
```
Comment (yamins81) [inline tfutils/base.py]: fuck that's an error sorry, will fix

File: tfutils/base.py [hunk 9]
```diff
@@ -593,17 +719,63 @@ def run_base(saver_params,
...
+
+
+def get_valid_targets_dict(validation_params, 
+                           model_func, model_kwargs,
+                           default_queue_params,
+                           cfg_final):
+    """NB: this function may modify validation_params"""
+    valid_targets_dict = OrderedDict()
+    queues = []
+    if validation_params is not None:
```
Comment (yamins81) [inline tfutils/base.py]: agreed will remove

File: tfutils/base.py [hunk 10]
```diff
@@ -593,17 +719,63 @@ def run_base(saver_params,
...
+        for vtarg in validation_params:
+            vdata_kwargs = copy.deepcopy(validation_params[vtarg]['data'])
```
Comment (yamins81) [inline tfutils/base.py]: agreed

Comment (yamins81) [inline tfutils/base.py]: Well, I cleaned it up in a way that's a little different that what you asked for.   This is because I actually want to modify validation_params itself as a side effect if needed, so that the correct full information is stored in the DB.

File: tfutils/base.py [hunk 11]
```diff
@@ -97,83 +98,131 @@ def __init__(self,
...
-            - force_fetch (bool, default: False)
-                Whether to fetch stored model from database even if its locally cached
             - *args, **kwargs
                 Additional arguments are passed onto base Saver class constructor
         """
 
-        SONified_params = SONify(params)
-        super(Saver, self).__init__(*args, **kwargs)
+        self.params = params
+        self.SONified_params = SONify(self.params)
```
Comment (qbilius) [inline tfutils/base.py]: I understood the reason but I think `sonify` is just as clear and much nicer.

File: tfutils/base.py (hunk 6 above)
Comment (qbilius) [inline tfutils/base.py]: OK

File: tfutils/base.py [hunk 12]
```diff
@@ -184,26 +233,35 @@ def load_from_db(self, query, cache_model=False, force_fetch=False):
...
 
-        count_recent = self.collfs_recent.find(query).count()
+        count_recent = collfs_recent.find(query).count()
         if count_recent > 0:  # get latest that matches query
-            ckpt_record_recent = self.coll_recent.find(query,
-                                                       sort=[('uploadDate', -1)])[0]
+            ckpt_record_recent = coll_recent.find(query,
+                                            sort=[('uploadDate', -1)])[0]
             # use the record with latest timestamp
             if ckpt_record is None or ckpt_record_recent['uploadDate'] > ckpt_record['uploadDate']:
```
Comment (qbilius) [inline tfutils/base.py]: if `ckpt_record` is None, this statement will raise an error since `ckpt_record['uploadDate']` is undefined.

File: tfutils/base.py (hunk 5 above)
Comment (qbilius) [inline tfutils/base.py]: Yeah, I've seen this a lot in dldata but laziness to type doesn't sound as a good enough excuse to rename variables. It's asking for troubles in my view.

File: tfutils/base.py [hunk 13]
```diff
@@ -742,15 +749,15 @@ def train_from_params(save_params,
...
+                           original_seed):
     """NB: this function may modify validation_params"""
     valid_targets_dict = OrderedDict()
     queues = []
-
     for vtarg in validation_params:
         if 'targets' not in validation_params[vtarg]:
             validation_params[vtarg]['targets'] = default_loss_params()
-        if 'func' not in validation_params[vtarg]['targets']:
+        if 'func' not in validatfion_params[vtarg]['targets']:
```
Comment (qbilius) [inline tfutils/base.py]: `validatfion_params` -> `validation_params`

File: tfutils/base.py (hunk 10 above)
Comment (qbilius) [inline tfutils/base.py]: Sure. I think it could be simplified even further (there are many variables that are used once, all named very similarly, so it is hard to follow) but let's deal with that later.

File: tfutils/base.py (hunk 12 above)
Comment (yamins81) [inline tfutils/base.py]: No, because the first clause, checking if ckpt_record is None, will, if true, abort the remainder of the clause from being checked.

File: tfutils/base.py (hunk 13 above)
Comment (yamins81) [inline tfutils/base.py]: Fixed -- there's something wrong with agent1 making it next to impossible to run the tests -- things are moving so slow.

File: tfutils/base.py (hunk 12 above)
Comment (qbilius) [inline tfutils/base.py]: Wow. I had it wrong for so many years! I've checked it now and you're right.

File: tfutils/base.py [hunk 14]
```diff
@@ -97,83 +98,131 @@ def __init__(self,
...
+                                                     port=self.load_port)
+        else:
+            self.load_conn = self.conn
+        self.load_collfs = gridfs.GridFS(self.load_conn[self.load_dbname], 
+                                             self.load_collname)
+        load_recent_name = '_'.join([self.load_dbname,
+                                     self.load_collname,
+                                     self.load_exp_id,
+                                     '__RECENT'])
+        self.load_collfs_recent = gridfs.GridFS(self.load_conn[load_recent_name])
```
Comment (yamins81) [inline tfutils/base.py]: Hm.  But then do you want me to do the save for self.collfs and self.collfs_recent?    It feels like these (and maybe the load_ versions) might be useful in other methods at some point.

File: tfutils/base.py (hunk 11 above)
Comment (yamins81) [inline tfutils/base.py]: ok changed

File: tfutils/base.py (hunk 8 above)
Comment (yamins81) [inline tfutils/base.py]: Something is going wrong with my editor causing TABs to be in places I did not intent them.  I'm going to have to check over the code line by line

File: tfutils/base.py (hunk 14 above)
Comment (qbilius) [inline tfutils/base.py]: Not sure what you mean by "do the save for self.collfs and self.collfs_recent". You mean should we have them defined here? I'd say yes. But load_ stuff I don't see useful, at least not yet.

Comment (yamins81) [inline tfutils/base.py]: Sorry, I meant, "do the same."    If we're going to construct self.collfs and self.collfs_recent in the initializer, let's do the same for the load_ stuff, since they're of the same "importance" and serve a similar function.  We shouldn't split up the initialization of the two sets of variables.

Comment (yamins81) [inline tfutils/base.py]: In other words, let's just leave them all where they are for now, in the __init__ method.
